### PR TITLE
hv: vcr: check guest cr3 before loading pdptrs

### PR DESCRIPTION
**Pull request by ssqre**
Check whether the address area pointed by the guest
cr3 is valid or not before loading pdptrs. Inject #GP(0)
to guest if there are any invalid cases.

Tracked-On: #3572
Signed-off-by: Jie Deng <jie.deng@intel.com>
Reviewed-by: Jason Chen CJ <jason.cj.chen@intel.com>
Acked-by: Eddie Dong <eddie.dong@intel.com>

### CODE REVIEW

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 1]
```diff
@@ -34,23 +34,52 @@
...
+	 */
+	if (copy_from_gpa(vcpu->vm, pdpte, get_pae_pdpt_addr(guest_cr3), sizeof(pdpte)) != 0) {
+		ret = -EFAULT;
+	} else {
+		/* Check if any of the PDPTEs sets both the P flag
+		 * and any reserved bit
+		 */
+		maxphyaddr = cpu_info->phys_bits;
+		/* reserved bits: 1~2, 5~8, maxphyaddr ~ 63 */
+		rsvd_bits_mask = (63 < maxphyaddr) ? 0 : (((1ULL << (63 - maxphyaddr + 1)) - 1) << maxphyaddr);
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/virtual_cr.c]: careful on MISRAC standard.

**Comment by HuiHuangShi** [comment]
28 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/include/arch/x86/cpu.h func:stac offset:494<br> reason:Static procedure is not explicitly called in code analysed. : stac<br><2> config:common filename:/hypervisor/include/arch/x86/cpu.h func:clac offset:499<br> reason:Static procedure is not explicitly called in code analysed. : clac<br><3> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:45<br> reason:Basic type declaration used. : int load_pdptrs<br><4> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:49<br> reason:Basic type declaration used. : int ret<br><5> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:52<br> reason:Basic type declaration used. : int maxphyaddr<br><6> config:sharing_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:63<br> reason:Pointer not checked for null before use. : cpu_info<br><7> config:part_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:63<br> reason:Pointer not checked for null before use. : cpu_info<br><8> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:63<br> reason:Use of mixed mode arithmetic. :  (int and unsigned char): maxphyaddr = cpu_info -> phys_bits<br><9> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Use of mixed mode arithmetic. :  (unsigned long and signed char): ( 1ULL << ( 63 - maxphyaddr + 1 ) ) - 1 ) << maxphyaddr<br><10> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Use of mixed mode arithmetic. :  (unsigned long and signed char): ( 1ULL << ( 63 - maxphyaddr + 1 ) ) - 1<br><11> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Use of mixed mode arithmetic. :  (unsigned char and signed char): ( 1ULL << ( 63 - maxphyaddr + 1 ) ) - 1<br><12> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Use of mixed mode arithmetic. :  (signed char and int): ( 1ULL << ( 63 - maxphyaddr<br><13> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Use of mixed mode arithmetic. :  (signed char and int): rsvd_bits_mask = ( 63 < maxphyaddr<br><14> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Type mismatch in ternary expression. :  (unsigned long and signed char)<br><15> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Implicit conversion of underlying type (MR). :  (signed char and int): rsvd_bits_mask = ( 63 < maxphyaddr<br><16> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Implicit conversion of underlying type (MR). :  (signed char and int): ( 1ULL << ( 63 - maxphyaddr<br><17> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Negative (or potentially negative) shift. : int : ( 63 - maxphyaddr + 1 )<br><18> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Value is not of appropriate type. :  (unsigned long and signed char): rsvd_bits_mask = (   63 <<br><19> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:65<br> reason:Use of shift operator on signed type. : << used with signed char: ( ( 1ULL << ( 63 - maxphyaddr + 1 ) ) - 1 )<br><20> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:66<br> reason:Literal value requires a U suffix.           : 0x00000000000001E6<br><21> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:66<br> reason:Use of bit operator on signed type. : |= used with int: 0x00000000000001E6<br><22> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:68<br> reason:Expression is not Boolean.<br><23> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:145<br> reason:Basic type declaration used. : int err_code<br><24> config:sharing_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:154<br> reason:Function return value potentially unused. : cr0_changed_bits<br><25> config:part_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:154<br> reason:Function return value potentially unused. : cr0_changed_bits<br><26> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr4 offset:307<br> reason:Basic type declaration used. : int err_code<br><27> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr4 offset:324<br> reason:No brackets to then/else.<br><28> config:common filename:/hypervisor/include/arch/x86/pgtable.h func:get_pgentry offset:252<br> reason:Static procedure is not explicitly called in code analysed. : get_pgentry<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: hypervisor/arch/x86/guest/virtual_cr.c (hunk 1 above)
Comment (HuiHuangShi) [inline hypervisor/arch/x86/guest/virtual_cr.c]: rsvd_bits_mask = (63 < maxphyaddr) ? 0UL : (((1UL << (63U - maxphyaddr + 1U)) - 1UL) << maxphyaddr);
maxphyaddr type shall be u32_t.

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 2]
```diff
@@ -277,23 +315,28 @@ static void vmx_write_cr4(struct acrn_vcpu *vcpu, uint64_t cr4)
...
 			if (((cr4 & CR4_PAE) != 0UL) && (is_paging_enabled(vcpu)) && (!is_long_mode(vcpu))) {
-				load_pdptrs(vcpu);
+				if (load_pdptrs(vcpu) != 0) {
+					err_code = -EFAULT;
+					vcpu_inject_gp(vcpu, 0U);
+				}
 			}
-
-			vcpu_make_request(vcpu, ACRN_REQUEST_EPT_FLUSH);
+			if (err_code == 0)
```
Comment (HuiHuangShi) [inline hypervisor/arch/x86/guest/virtual_cr.c]: if (err_code == 0) {
   ....
}

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 3]
```diff
@@ -113,6 +142,8 @@ static bool is_cr0_write_valid(struct acrn_vcpu *vcpu, uint64_t cr0)
  */
 static void vmx_write_cr0(struct acrn_vcpu *vcpu, uint64_t cr0)
 {
+	int err_code = 0;
```
Comment (HuiHuangShi) [inline hypervisor/arch/x86/guest/virtual_cr.c]: int32_t err_code = 0;

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 4]
```diff
@@ -268,6 +304,8 @@ static bool is_cr4_write_valid(struct acrn_vcpu *vcpu, uint64_t cr4)
  */
 static void vmx_write_cr4(struct acrn_vcpu *vcpu, uint64_t cr4)
 {
+	int err_code = 0;
```
Comment (HuiHuangShi) [inline hypervisor/arch/x86/guest/virtual_cr.c]: int32_t

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 5]
```diff
@@ -34,23 +34,52 @@
 				CR4_OSXMMEXCPT | CR4_SMAP | CR4_PKE |        \
 				CR4_SMXE | CR4_UMIP)
 
+/* PAE PDPTE bits 1 ~ 2, 5 ~ 8 are always reserved */
+#define PAE_PDPTE_FIXED_RESVD_BITS	0x00000000000001E6
```
Comment (HuiHuangShi) [inline hypervisor/arch/x86/guest/virtual_cr.c]: add UL

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 6]
```diff
@@ -34,23 +34,52 @@
...
+	} else {
+		/* Check if any of the PDPTEs sets both the P flag
+		 * and any reserved bit
+		 */
+		maxphyaddr = cpu_info->phys_bits;
+		/* reserved bits: 1~2, 5~8, maxphyaddr ~ 63 */
+		rsvd_bits_mask = (63 < maxphyaddr) ? 0 : (((1ULL << (63 - maxphyaddr + 1)) - 1) << maxphyaddr);
+		rsvd_bits_mask |= PAE_PDPTE_FIXED_RESVD_BITS;
+		for (i = 0; i < 4; i++) {
+			if ((pdpte[i] & PAGE_PRESENT) && (pdpte[i] & rsvd_bits_mask)) {
```
Comment (HuiHuangShi) [inline hypervisor/arch/x86/guest/virtual_cr.c]: if (((pdpte[i] & PAGE_PRESENT) != 0) && ((pdpte[i] & rsvd_bits_mask) != 0)) {

**Comment by acrnsi** [comment]
OK to verify

**Comment by HuiHuangShi** [comment]
8 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/include/arch/x86/cpu.h func:stac offset:494<br> reason:Static procedure is not explicitly called in code analysed. : stac<br><2> config:common filename:/hypervisor/include/arch/x86/cpu.h func:clac offset:499<br> reason:Static procedure is not explicitly called in code analysed. : clac<br><3> config:sharing_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:64<br> reason:Pointer not checked for null before use. : cpu_info<br><4> config:part_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:64<br> reason:Pointer not checked for null before use. : cpu_info<br><5> config:common filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:69<br> reason:Literal value requires a U suffix.           : 0<br><6> config:sharing_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:155<br> reason:Function return value potentially unused. : cr0_changed_bits<br><7> config:part_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:155<br> reason:Function return value potentially unused. : cr0_changed_bits<br><8> config:common filename:/hypervisor/include/arch/x86/pgtable.h func:get_pgentry offset:252<br> reason:Static procedure is not explicitly called in code analysed. : get_pgentry<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 7]
```diff
@@ -34,23 +34,53 @@
...
+	} else {
+		/* Check if any of the PDPTEs sets both the P flag
+		 * and any reserved bit
+		 */
+		maxphyaddr = cpu_info->phys_bits;
+		/* reserved bits: 1~2, 5~8, maxphyaddr ~ 63 */
+		rsvd_bits_mask = (63U < maxphyaddr) ? 0UL : (((1UL << (63U - maxphyaddr + 1U)) - 1UL) << maxphyaddr);
+		rsvd_bits_mask |= PAE_PDPTE_FIXED_RESVD_BITS;
+		for (i = 0; i < 4; i++) {
+			if (((pdpte[i] & PAGE_PRESENT) != 0) && ((pdpte[i] & rsvd_bits_mask) != 0)) {
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/virtual_cr.c]: 0U

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 8]
```diff
@@ -113,6 +143,8 @@ static bool is_cr0_write_valid(struct acrn_vcpu *vcpu, uint64_t cr0)
  */
 static void vmx_write_cr0(struct acrn_vcpu *vcpu, uint64_t cr0)
 {
+	int32_t err_code = 0;
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/virtual_cr.c]: why err_code is signed int?

File: hypervisor/arch/x86/guest/virtual_cr.c [hunk 9]
```diff
@@ -140,7 +172,10 @@ static void vmx_write_cr0(struct acrn_vcpu *vcpu, uint64_t cr0)
 				vcpu_set_efer(vcpu, vcpu_get_efer(vcpu) | MSR_IA32_EFER_LMA_BIT);
 			} else if (is_pae(vcpu)) {
 				/* enabled PAE from paging disabled */
-				load_pdptrs(vcpu);
+				if (load_pdptrs(vcpu) != 0) {
+					err_code = -EFAULT;
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/virtual_cr.c]: -EFAULT is not used later on.
err_code ? or bool err_found? or load_failure? or failure? bool is better

**Comment by acrnsi** [comment]
hi ,@ssqre ,thanks for your pr . your Tracked-On info  is error. It should be "Tracked-On" rather than "Tracked-on" in your commits.

File: hypervisor/arch/x86/guest/virtual_cr.c (hunk 9 above)
Comment (ssqre) [reply hypervisor/arch/x86/guest/virtual_cr.c]: ok. thanks

File: hypervisor/arch/x86/guest/virtual_cr.c (hunk 7 above)
Comment (ssqre) [reply hypervisor/arch/x86/guest/virtual_cr.c]: Got it. Thanks.

**Comment by HuiHuangShi** [comment]
7 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/include/arch/x86/cpu.h func:stac offset:494<br> reason:Static procedure is not explicitly called in code analysed. : stac<br><2> config:common filename:/hypervisor/include/arch/x86/cpu.h func:clac offset:499<br> reason:Static procedure is not explicitly called in code analysed. : clac<br><3> config:sharing_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:64<br> reason:Pointer not checked for null before use. : cpu_info<br><4> config:part_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:load_pdptrs offset:64<br> reason:Pointer not checked for null before use. : cpu_info<br><5> config:sharing_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:155<br> reason:Function return value potentially unused. : cr0_changed_bits<br><6> config:part_config filename:/hypervisor/arch/x86/guest/virtual_cr.c func:vmx_write_cr0 offset:155<br> reason:Function return value potentially unused. : cr0_changed_bits<br><7> config:common filename:/hypervisor/include/arch/x86/pgtable.h func:get_pgentry offset:252<br> reason:Static procedure is not explicitly called in code analysed. : get_pgentry<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Pull request merged**